In [1]:
%matplotlib inline
import pandas as pd
import numpy as np
import seaborn as sns 
import matplotlib.pyplot as plt
from scipy import stats
from ast import literal_eval
from sklearn.feature_extraction.text import CountVectorizer,TfidfVectorizer
from sklearn.metrics.pairwise import linear_kernel,cosine_similarity
from nltk.stem.snowball import SnowballStemmer
from nltk.stem.wordnet import WordNetLemmatizer
from nltk.corpus import wordnet
import warnings;warnings.simplefilter('ignore')
from IPython.display import HTML,Image
pd.set_option('display.max_colwidth', 150)

## Content
Simple Recommender :
1.1 IMDB Top 250: Top 250 Movies based on calculated IMDB ratings
1.2 Recommendation by Genre: Top Movies for every genre in database 

Content Based Recommender:
2.1 Movie Description Based: Recommend movies based on overview and tagline
2.2 Metadata based: Recommend movies based on Cast, Director and keywords of movie

Collaborative Filtering using Surprise:
The personalised recommender based on users Past history of ratings and similarity between his and other users rating history

Hybrid Recommender:
Recommeder system based on collaberation of Contend based recommender and collaberative filtering, It leverages features of both recommender for better recommendations

In [2]:
df = pd.read_csv('movies_metadata.csv')

In [3]:
df.head()

adult  \
0  False   
1  False   
2  False   
3  False   
4  False   

                                                                                                                                   belongs_to_collection  \
0  {'id': 10194, 'name': 'Toy Story Collection', 'poster_path': '/7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg', 'backdrop_path': '/9FBwqcd9IRruEDUrTdcaafOMKUq.jpg'}   
1                                                                                                                                                    NaN   
2  {'id': 119050, 'name': 'Grumpy Old Men Collection', 'poster_path': '/nLvUdqgPgm3F85NMCii9gVFUcet.jpg', 'backdrop_path': '/hypTnLot2z8wpFS7qwsQHW1u...   
3                                                                                                                                                    NaN   
4  {'id': 96871, 'name': 'Father of the Bride Collection', 'poster_path': '/nts4iOmNnq7GNicycMJ9pSAn204.jpg', 'backdrop_path': '/7qwE57OVZmMJChBpLEbJ...   

     budget  \
0  30000000   
1  65000000   
2         0   
3  16000000   
4         0   

                                                                                              genres  \
0   [{'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}]   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, 'name': 'Fantasy'}, {'id': 10751, 'name': 'Family'}]   
2                                   [{'id': 10749, 'name': 'Romance'}, {'id': 35, 'name': 'Comedy'}]   
3      [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'name': 'Drama'}, {'id': 10749, 'name': 'Romance'}]   
4                                                                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                                                                                                                                overview  \
0  Led by Woody, Andy's toys live happily in his room until Andy's birthday brings Buzz Lightyear onto the scene. Afraid of losing his place in Andy'...   
1  When siblings Judy and Peter discover an enchanted board game that opens the door to a magical world, they unwittingly invite Alan -- an adult who...   
2  A family wedding reignites the ancient feud between next-door neighbors and fishing buddies John and Max. Meanwhile, a sultry Italian divorcée ope...   
3  Cheated on, mistreated and stepped on, the women are holding their breath, waiting for the elusive "good man" to break a string of less-than-stell...   
4  Just when George Banks has recovered from his daughter's wedding, he receives the news that she's pregnant ... and that George's wife, Nina, is ex...   

   ... release_date      revenue runtime  \
0  ...   1995-10-30  373554033.0    81.0   
1  ...   1995-12-15  262797249.0   104.0   
2  ...   1995-12-22          0.0   101.0   
3  ...   1995-12-22   81452156.0   127.0   
4  ...   1995-02-10   76578911.0   106.0   

                                                                    spoken_languages  \
0                                           [{'iso_639_1': 'en', 'name': 'English'}]   
1  [{'iso_639_1': 'en', 'name': 'English'}, {'iso_639_1': 'fr', 'name': 'Français'}]   
2                                           [{'iso_639_1': 'en', 'name': 'English'}]   
3                                          

In [4]:
df.describe()

revenue       runtime  vote_average    vote_count
count  4.546000e+04  45203.000000  45460.000000  45460.000000
mean   1.120935e+07     94.128199      5.618207    109.897338
std    6.433225e+07     38.407810      1.924216    491.310374
min    0.000000e+00      0.000000      0.000000      0.000000
25%    0.000000e+00     85.000000      5.000000      3.000000
50%    0.000000e+00     95.000000      6.000000     10.000000
75%    0.000000e+00    107.000000      6.800000     34.000000
max    2.787965e+09   1256.000000     10.000000  14075.000000

In [5]:
df.columns

Index(['adult', 'belongs_to_collection', 'budget', 'genres', 'homepage', 'id',
       'imdb_id', 'original_language', 'original_title', 'overview',
       'popularity', 'poster_path', 'production_companies',
       'production_countries', 'release_date', 'revenue', 'runtime',
       'spoken_languages', 'status', 'tagline', 'title', 'video',
       'vote_average', 'vote_count'],
      dtype='object')

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 24 columns):
adult                    45466 non-null object
belongs_to_collection    4494 non-null object
budget                   45466 non-null object
genres                   45466 non-null object
homepage                 7782 non-null object
id                       45466 non-null object
imdb_id                  45449 non-null object
original_language        45455 non-null object
original_title           45466 non-null object
overview                 44512 non-null object
popularity               45461 non-null object
poster_path              45080 non-null object
production_companies     45463 non-null object
production_countries     45463 non-null object
release_date             45379 non-null object
revenue                  45460 non-null float64
runtime                  45203 non-null float64
spoken_languages         45460 non-null object
status                   45379 non-null objec

In [7]:
df['genres'] = df['genres'].fillna('[]').apply(literal_eval).apply(lambda x : [i['name'] for i in x] if isinstance(x,list) else [])

# 1. Simple Recommender
This is a very basic model based on movies popularity and critical acclaimation, this model does not give personalised recommendation

I will use IMDB's weighted rating formula to construct my chart. Mathematically, it is represented as follows:
Weighted Rating (WR) = ((v/(v+m)).R)+((m/(v+m)).C)
where,
v is the number of votes for the movie
m is the minimum votes required to be listed in the chart
R is the average rating of the movie
C is the mean vote across the whole report

In [8]:
base_poster_url = 'http://image.tmdb.org/t/p/w185/'
df['poster_path'] = "<img src='" + base_poster_url + df['poster_path'] + "' style='height:100px;'>"

In [9]:
df['vote_count'] = df[df['vote_count'].notnull()]['vote_count'].astype('int')

In [10]:
df['vote_average'] = df[df['vote_average'].notnull()]['vote_average'].astype('int')

In [11]:
c = df['vote_average'].mean()

In [12]:
m = df['vote_count'].quantile(0.95)

In [13]:
df['year'] = pd.to_datetime(df['release_date'],errors='coerce').apply(lambda x : str(x).split('-')[0] if x!= np.nan else np.nan)

In [14]:
qualified = df[(df['vote_count'] >= m) & (df['vote_average'].notnull()) & (df['vote_count'].notnull())][['poster_path','title','year','popularity','vote_count','vote_average','genres']]
HTML(qualified.head(2).to_html(escape=False))

In [15]:
qualified.shape

(2274, 7)

There are total 2274 movies that are qualified to be considered for the recommender charts.
Lets calculated weighted rating according to mentioned IMDB's formulae

In [16]:
qualified['vote_count'] = qualified['vote_count'].astype('int')
qualified['vote_average'] = qualified['vote_average'].astype('int')

In [17]:
def weighted_rating(x):
    v = x['vote_count']
    R = x['vote_average']
    return(v/(v+m) * R) + (m/(m+v) * c)


In [18]:
qualified['wr'] = qualified.apply(weighted_rating,axis=1)
HTML(qualified.head(2).to_html(escape=False))

# 1.1 IMDB TOP 250

In [19]:
qualified.sort_values('wr',axis=0,ascending=False).head(250)
HTML(qualified.sort_values('wr',axis=0,ascending=False).head(250).to_html(escape=False))

# 1.2 Recommendation by genre

In [20]:
s= df.apply(lambda x : pd.Series(x['genres']),axis=1).stack().reset_index(level=1,drop=True)
s.name = 'genres'

In [21]:
gen_df = df.drop('genres',axis=1).join(s)

In [22]:
gen_df.head()

adult  \
0  False   
0  False   
0  False   
1  False   
1  False   

                                                                                                                                   belongs_to_collection  \
0  {'id': 10194, 'name': 'Toy Story Collection', 'poster_path': '/7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg', 'backdrop_path': '/9FBwqcd9IRruEDUrTdcaafOMKUq.jpg'}   
0  {'id': 10194, 'name': 'Toy Story Collection', 'poster_path': '/7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg', 'backdrop_path': '/9FBwqcd9IRruEDUrTdcaafOMKUq.jpg'}   
0  {'id': 10194, 'name': 'Toy Story Collection', 'poster_path': '/7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg', 'backdrop_path': '/9FBwqcd9IRruEDUrTdcaafOMKUq.jpg'}   
1                                                                                                                                                    NaN   
1                                                                                                                                                    NaN   

     budget                              homepage    id    imdb_id  \
0  30000000  http://toystory.disney.com/toy-story   862  tt0114709   
0  30000000  http://toystory.disney.com/toy-story   862  tt0114709   
0  30000000  http://toystory.disney.com/toy-story   862  tt0114709   
1  65000000                                   NaN  8844  tt0113497   
1  65000000                                   NaN  8844  tt0113497   

  original_language original_title  \
0                en      Toy Story   
0                en      Toy Story   
0                en      Toy Story   
1                en        Jumanji   
1                en        Jumanji   

                                                                                                                                                overview  \
0  Led by Woody, Andy's toys live happily in his room until Andy's birthday brings Buzz Lightyear onto the scene. Afraid of losing his place in Andy'...   
0  Led by Woody, Andy's toys live happily in his room until Andy's birthday brings Buzz Lightyear onto the scene. Afraid of losing his place in Andy'...   
0  Led by Woody, Andy's toys live happily in his room until Andy's birthday brings Buzz Lightyear onto the scene. Afraid of losing his place in Andy'...   
1  When siblings Judy and Peter discover an enchanted board game that opens the door to a magical world, they unwittingly invite Alan -- an adult who...   
1  When siblings Judy and Peter discover an enchanted board game that opens the door to a magical world, they unwittingly invite Alan -- an adult who...   

  popularity  ... runtime  \
0    21.9469  ...    81.0   
0    21.9469  ...    81.0   
0    21.9469  ...    81.0   
1    17.0155  ...   104.0   
1    17.0155  ...   104.0   

                                                                    spoken_languages  \
0                                           [{'iso_639_1': 'en', 'name': 'English'}]   
0                                           [{'iso_639_1': 'en', 'name': 'English'}]   
0                                           [{'iso_639_1': 'en', 'name': 'English'}]   
1  [{'iso_639_1': 'en', 'name': 'English'}, {'iso_639_1': 'fr', 'name': 'Français'}]   
1  [{'iso_639_1': 'en', 'name': 'English'}, {'iso_639_1': 'fr', 'name': 'Français'}]   

     status                                    tagline      title  video  \
0  Released                                        NaN  Toy Story  False   
0  Released                                        NaN  Toy Story  False   
0  Released                                        NaN  Toy Story  False   
1  Released  Roll the dice and unleash the excitement!    Jumanji  False   
1  Released  Roll the dice and unleash the excitement!    Jumanji  False   

  vote_average vote_count  year     genres  
0          7.0     5415.0  1995  Animation  
0          7.0     5415.0  1995     Comedy  
0          7.0     5415.0  1995     Family  
1          6.0     2413.0  1995  Adventure  
1          6.0     2

In [23]:
def build_chart(genre,percentile=0.85):
    df=gen_df[gen_df['genres']== genre]
    vote_counts = df[df['vote_count'].notnull()]['vote_count'].astype('int')
    vote_averages = df[df['vote_average'].notnull()]['vote_average'].astype('int')
    C = vote_averages.mean()
    m = vote_counts.quantile(percentile)
    
    qualified = df[(df['vote_count'] >= m) & (df['vote_average'].notnull()) & (df['vote_count'].notnull())][['poster_path','title','year','popularity','vote_count','vote_average','genres']]
    qualified['vote_count'] = qualified['vote_count'].astype('int')
    qualified['vote_average'] = qualified['vote_average'].astype('int')
    
    qualified['wr'] = qualified.apply(lambda x: (x['vote_count']/(x['vote_count']+m) * x['vote_average']) + (m/(m+x['vote_count']) * C), axis=1)
    qualified = qualified.sort_values('wr',ascending=False).head(250)
    return qualified
    
    

In [24]:
build_chart('Romance',percentile=0.85)
HTML(build_chart('Romance',percentile=0.85).head(15).to_html(escape=False))

As we move Percentile value, the movies to qualify for recomendation get selected or dropped, Higher the percentile, higher number of votes required to qualify.

In [25]:
build_chart('Thriller',percentile=0.85).head(15)
HTML(build_chart('Thriller',percentile=0.85).head(15).to_html(escape=False))

Well these are quite good thriller  movies, but Pulp Fiction is still my favourite thriller  movie, whatever this list says
The above recommender is not personalised and it will only generate recommendation by genre, irrespective of persons own interest
## 2. Content Based Recommender
To personalise our recommendations more, I am going to build an engine that computes similarity between movies based on certain metrics and suggests movies that are most similar to a particular movie that a user liked. Since we will be using movie metadata (or content) to build this engine, this also known as Content Based Filtering.

Even though content based recommender is still not personalised recommender, we gonna use it letter for hybrid recommender

I will build 2 content based Recommender based on:

1.Movie Overview and tagline

2.Movie cast,director, genre,keywords

For upcoming recommenders, we gonna need sparse matrix formed using Cosine similarities/linear similarities, so for computation conveninece we will subset of dataset

In [26]:
links_small = pd.read_csv('links_small.csv')

In [27]:
links_small.head()

movieId  imdbId   tmdbId
0        1  114709    862.0
1        2  113497   8844.0
2        3  113228  15602.0
3        4  114885  31357.0
4        5  113041  11862.0

In [28]:
links_small.shape

(9125, 3)

In [29]:
links_small = links_small[links_small['tmdbId'].notnull()]['tmdbId'].astype('int')

In [30]:
df = df.drop([19730, 29503, 35587])

In [31]:
df['id'] = df['id'].astype('int')

In [32]:
smd = df[df['id'].isin(links_small)]

In [33]:
smd.shape

(9099, 25)

In [34]:
smd.head()

adult  \
0  False   
1  False   
2  False   
3  False   
4  False   

                                                                                                                                   belongs_to_collection  \
0  {'id': 10194, 'name': 'Toy Story Collection', 'poster_path': '/7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg', 'backdrop_path': '/9FBwqcd9IRruEDUrTdcaafOMKUq.jpg'}   
1                                                                                                                                                    NaN   
2  {'id': 119050, 'name': 'Grumpy Old Men Collection', 'poster_path': '/nLvUdqgPgm3F85NMCii9gVFUcet.jpg', 'backdrop_path': '/hypTnLot2z8wpFS7qwsQHW1u...   
3                                                                                                                                                    NaN   
4  {'id': 96871, 'name': 'Father of the Bride Collection', 'poster_path': '/nts4iOmNnq7GNicycMJ9pSAn204.jpg', 'backdrop_path': '/7qwE57OVZmMJChBpLEbJ...   

     budget                        genres  \
0  30000000   [Animation, Comedy, Family]   
1  65000000  [Adventure, Fantasy, Family]   
2         0             [Romance, Comedy]   
3  16000000      [Comedy, Drama, Romance]   
4         0                      [Comedy]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                                                                                                                                overview  \
0  Led by Woody, Andy's toys live happily in his room until Andy's birthday brings Buzz Lightyear onto the scene. Afraid of losing his place in Andy'...   
1  When siblings Judy and Peter discover an enchanted board game that opens the door to a magical world, they unwittingly invite Alan -- an adult who...   
2  A family wedding reignites the ancient feud between next-door neighbors and fishing buddies John and Max. Meanwhile, a sultry Italian divorcée ope...   
3  Cheated on, mistreated and stepped on, the women are holding their breath, waiting for the elusive "good man" to break a string of less-than-stell...   
4  Just when George Banks has recovered from his daughter's wedding, he receives the news that she's pregnant ... and that George's wife, Nina, is ex...   

   ...      revenue runtime  \
0  ...  373554033.0    81.0   
1  ...  262797249.0   104.0   
2  ...          0.0   101.0   
3  ...   81452156.0   127.0   
4  ...   76578911.0   106.0   

                                                                    spoken_languages  \
0                                           [{'iso_639_1': 'en', 'name': 'English'}]   
1  [{'iso_639_1': 'en', 'name': 'English'}, {'iso_639_1': 'fr', 'name': 'Français'}]   
2                                           [{'iso_639_1': 'en', 'name': 'English'}]   
3                                           [{'iso_639_1': 'en', 'name': 'English'}]   
4                                           [{'iso_639_1': 'en', 'name': 'English'}]   

     status  \
0  Released   
1  Released   
2  Released   
3  Released   
4  Released   

                                                                          tagline  \
0                                                                             NaN   
1                                       Roll the dice and unleash the excitement!   
2                            Still Yelling.

 So we will work with 9099 movies

## 2.1 Movie description based recommender
Based on overview and tagline

In [35]:
smd['tagline'] = smd['tagline'].fillna('')
smd['description'] = smd['overview'] + smd['tagline']
smd['description'] = smd['description'].fillna('')

In [36]:
tf = TfidfVectorizer(analyzer='word',ngram_range=(1,2),min_df=0,stop_words='english')

In [37]:
tfid_matrix = tf.fit_transform(smd['description'])

# Cosine Similarity
We Will use cosine similarity to calculate numeric quantity that denote similarity between 2 movies. More info about cosine similarities . https://scikit-learn.org/stable/modules/metrics.html

In [38]:
cosin_sim= linear_kernel(tfid_matrix,tfid_matrix)

In [39]:
cosin_sim[0]

array([1.        , 0.00680476, 0.        , ..., 0.        , 0.00344913,
       0.        ])

In [40]:
smd = smd.reset_index()
titles = smd['title']
indices = pd.Series(smd.index,index=smd['title'])

In [41]:
def get_recommendations(title):
    idx = indices[title]
    sim_scores = list(enumerate(cosin_sim[idx]))
    sim_scores = sorted(sim_scores,key = lambda x : x[1] ,reverse= True)
    sim_scores = sim_scores[1:31]
    movies_indices = [i[0] for i in sim_scores]
    return smd.iloc[movies_indices][['poster_path','title','year','description']]
    

In [42]:
get_recommendations('The Godfather')
HTML(get_recommendations('The Godfather').to_html(escape=False))

In [43]:
get_recommendations('The Dark Knight')
HTML(get_recommendations('The Dark Knight').to_html(escape=False))

 As we can see the recommeder is recommending the movie whose plotline is similar to 'The Dark Knight', which is no surprise that it is showing most of the batman movies

In [44]:
get_recommendations('Q & A')
HTML(get_recommendations('Q & A').to_html(escape=False))

# 2.2 Metadata Based Recommender¶
It will be based on cast, crew, genre, keywords

In [45]:
credits =pd.read_csv('credits.csv')
keywords = pd.read_csv('keywords.csv')

In [46]:
keywords['id'] = keywords['id'].astype('int')
credits['id'] = credits['id'].astype('int')
df['id'] = df['id'].astype('int')

In [47]:
df.shape

(45463, 25)

In [48]:
df = df.merge(credits,on='id')
df = df.merge(keywords,on='id')

In [49]:
smd = df[df['id'].isin(links_small)]

In [50]:
smd.shape

(9219, 28)

In [51]:
smd['cast'] = smd['cast'].apply(literal_eval)
smd['crew'] = smd['crew'].apply(literal_eval)
smd['keywords'] = smd['keywords'].apply(literal_eval)
smd['cast_size'] = smd['cast'].apply(lambda x : len(x))
smd['crew_size'] = smd['crew'].apply(lambda x : len(x))

In Cast column, there are total 10-15 actors are present for each movie, but we need only top 3-4 main actors who have worked in movie.
Now we will get director

In [52]:
def get_director(x):
    for i in x:
        if i['job'] == 'Director':
            return i['name']
    return np.nan        

In [53]:
smd['director'] = smd['crew'].apply(get_director)

In [54]:
smd['cast'] = smd['cast'].apply(lambda x : [i['name'] for i in x]if isinstance (x,list) else [])
smd['cast'] = smd['cast'].apply(lambda x : x[:3] if len(x) >= 3 else x)

In [55]:
smd['cast'] = smd['cast'].apply(lambda x : [(str.lower(i.replace(" ","")) for i in x)])

In [56]:
smd['director'] = smd['director'].astype('str').apply(lambda x :str.lower(x.replace(" ","")))
smd['director'] = smd['director'].apply(lambda x : [x,x, x])

In [57]:
k = smd.apply(lambda x : pd.Series(x['keywords']),axis=1).stack().reset_index(level=1,drop=True)
k.name = 'keyword'

In [58]:
k.head()

0       {'id': 931, 'name': 'jealousy'}
0           {'id': 4290, 'name': 'toy'}
0           {'id': 5202, 'name': 'boy'}
0    {'id': 6054, 'name': 'friendship'}
0       {'id': 9713, 'name': 'friends'}
Name: keyword, dtype: object

In [59]:
k = k.value_counts()
k[:5]

{'id': 10183, 'name': 'independent film'}         610
{'id': 187056, 'name': 'woman director'}          550
{'id': 9826, 'name': 'murder'}                    399
{'id': 179431, 'name': 'duringcreditsstinger'}    327
{'id': 818, 'name': 'based on novel'}             318
Name: keyword, dtype: int64

In [60]:
k = k[k>1]

In [61]:
k.head()

{'id': 10183, 'name': 'independent film'}         610
{'id': 187056, 'name': 'woman director'}          550
{'id': 9826, 'name': 'murder'}                    399
{'id': 179431, 'name': 'duringcreditsstinger'}    327
{'id': 818, 'name': 'based on novel'}             318
Name: keyword, dtype: int64

 Now using Snowball stemmer we will stem keywords

In [62]:
stemmer = SnowballStemmer('english')
stemmer.stem('dogs')

'dog'

In [63]:
def filter_words(x):
    words = []
    for i in x :
        if i in k:
            words.append(i)
    return words

In [64]:
smd['keywords'] = smd['keywords'].apply(lambda x : [i['name'] for i in x] if isinstance (x,list) else [])

In [65]:
smd['keywords'] = smd['keywords'].apply(filter_words)
smd['keywords'] = smd['keywords'].apply(lambda x : [stemmer.stem(i) for i in x])
smd['keywords'] = smd['keywords'].apply(lambda x : [str.lower(i.replace(" ","")) for i in x])

In [66]:
smd['soup'] = smd['keywords'] + smd['director'] + smd['cast'] + smd['genres']
smd['soup'] = smd['soup'].astype('str')
smd['soup'] = smd['soup'].apply(lambda x : ''.join(x))

In [67]:
count = CountVectorizer(analyzer = 'word',min_df=0,ngram_range=(1,2),stop_words='english')
count_matrix = count.fit_transform(smd['soup'])

In [68]:
cosin_sim = linear_kernel(count_matrix,count_matrix)

In [69]:
smd = smd.reset_index()
titles = smd['title']
indices = pd.Series(smd.index , index= smd['title'])

In [70]:
def get_recommendations(title):
    idx = indices[title]
    sim_scores = list(enumerate(cosin_sim[idx]))
    sim_scores = sorted(sim_scores,key = lambda x : x[1] ,reverse= True)
    sim_scores = sim_scores[1:31]
    movies_indices = [i[0] for i in sim_scores]
    j= smd.iloc[movies_indices][['poster_path','title','year','cast','director']]
    j['director'] = j['director'].apply(lambda x:x[0])
    return j

As Our Cosine function is changed, lets see what recommendation we get now

In [71]:
get_recommendations('The Dark Knight')
HTML(get_recommendations('The Dark Knight').to_html(escape=False))

In [72]:
get_recommendations('Mean Girls')
HTML(get_recommendations('Mean Girls').to_html(escape=False))

# 2.3 improved recommendations 

In [73]:
def improved_recommendations(title):
    idx = indices[title]
    sim_scores = list(enumerate(cosin_sim[idx]))
    sim_scores = sorted(sim_scores,key=lambda x : x[1],reverse=True)
    sim_scores = sim_scores[1:26]
    movies_indices = [i[0] for i in sim_scores]
    
    movies = smd.iloc[movies_indices][['poster_path','title','vote_count','vote_average','year']]
    vote_counts = movies[movies['vote_count'].notnull()]['vote_count'].astype('int')
    vote_averages = movies[movies['vote_average'].notnull()]['vote_average'].astype('int')
    C = vote_averages.mean()
    m = vote_counts.quantile(0.60)
    qualified = movies[(movies['vote_count'] >= m) & (movies['vote_count'].notnull()) & (movies['vote_average'].notnull())]
    qualified['vote_count'] = qualified['vote_count'].astype('int')
    qualified['vote_average'] = qualified['vote_average'].astype('int')
    qualified['wr'] = qualified.apply(weighted_rating, axis=1)
    qualified = qualified.sort_values('wr', ascending=False).head(10)
    return qualified

In [74]:
improved_recommendations('The Dark Knight').head(10)
HTML(improved_recommendations('The Dark Knight').to_html(escape=False))

# 3. Collaborative Filtering
The above recommendation models has one limitation, It is not user centric, It will only recommend the movies which are close to the given movies. I will use Surprise library for more user centric recommendations

In [75]:
from surprise import SVD ,Reader,Dataset,evaluate

In [76]:
reader = Reader()

In [77]:
ratings = pd.read_csv('ratings_small.csv')

In [78]:
ratings.head()

userId  movieId  rating   timestamp
0       1       31     2.5  1260759144
1       1     1029     3.0  1260759179
2       1     1061     3.0  1260759182
3       1     1129     2.0  1260759185
4       1     1172     4.0  1260759205

In [79]:
data= Dataset.load_from_df(ratings[['userId','movieId','rating']],reader)

In [80]:
data.split(n_folds=5)

In [81]:
svd = SVD()
evaluate(svd,data,measures=['RMSE','MAE'])

Evaluating RMSE, MAE of algorithm SVD.

------------
Fold 1
RMSE: 0.8954
MAE:  0.6899
------------
Fold 2
RMSE: 0.9018
MAE:  0.6923
------------
Fold 3
RMSE: 0.9008
MAE:  0.6913
------------
Fold 4
RMSE: 0.8942
MAE:  0.6878
------------
Fold 5
RMSE: 0.8904
MAE:  0.6888
------------
------------
Mean RMSE: 0.8965
Mean MAE : 0.6900
------------
------------


CaseInsensitiveDefaultDict(list,
                           {'rmse': [0.8954385982162387,
                             0.9018135020707515,
                             0.9008467774689152,
                             0.8941614983332761,
                             0.8904498725968937],
                            'mae': [0.6898542646156821,
                             0.6923200416316841,
                             0.6912794457578506,
                             0.6878368018938282,
                             0.6888287700368798]})

In [82]:
trainset = data.build_full_trainset()

In [83]:
svd.train(trainset)

In [84]:
ratings[ratings['userId'] == 1]

userId  movieId  rating   timestamp
0        1       31     2.5  1260759144
1        1     1029     3.0  1260759179
2        1     1061     3.0  1260759182
3        1     1129     2.0  1260759185
4        1     1172     4.0  1260759205
5        1     1263     2.0  1260759151
6        1     1287     2.0  1260759187
7        1     1293     2.0  1260759148
8        1     1339     3.5  1260759125
9        1     1343     2.0  1260759131
10       1     1371     2.5  1260759135
11       1     1405     1.0  1260759203
12       1     1953     4.0  1260759191
13       1     2105     4.0  1260759139
14       1     2150     3.0  1260759194
15       1     2193     2.0  1260759198
16       1     2294     2.0  1260759108
17       1     2455     2.5  1260759113
18       1     2968     1.0  1260759200
19       1     3671     3.0  1260759117

In [85]:
svd.predict(1,302,3)

Prediction(uid=1, iid=302, r_ui=3, est=2.717807478839709, details={'was_impossible': False})

 The model predict rating given by userid 1 to movie id 33 is 2.60. This is purely on how other user have given rating to the movie and how their ratings are similar to userid1

## 4. Hybrid Recommender
We will collaborate the content based(Metadata based) and collaborative filtering recommenders

In [86]:
def convert_int(x):
    try:
        return int(x)
    except:
        return np.nan

In [87]:
id_map = pd.read_csv('links_small.csv')[['movieId','tmdbId']]

In [88]:
id_map['tmdbId'] = id_map['tmdbId'].apply(convert_int)

In [89]:
id_map.columns = ['movieId','id']

In [90]:
id_map= id_map.merge(smd[['title','id']],on='id').set_index('title')

In [91]:
indices_map = id_map.set_index('id')

In [92]:
def hybrid(userId,title):
    idx = indices[title]
    tmdbId = id_map.loc[title]['id']
    movie_id = id_map.loc[title]['movieId']
    
    sim_scores = list(enumerate(cosin_sim[int(idx)]))
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    sim_scores = sim_scores[1:26]
    movie_indices = [i[0] for i in sim_scores]
    
    movies = smd.iloc[movie_indices][['poster_path','title', 'vote_count', 'vote_average', 'year', 'id']]
    movies['est'] = movies['id'].apply(lambda x: svd.predict(userId, indices_map.loc[x]['movieId']).est)
    movies = movies.sort_values('est', ascending=False)
    return movies.head(10)

In [93]:
hybrid(1,'Avatar')
HTML(hybrid(1,'Avatar').to_html(escape=False))